nautobot/core/api/utils.py

```diff
@@ -353,3 +358,162 @@ def return_nested_serializer_data_based_on_depth(serializer, depth, obj, obj_rel
             return field_class(
                 obj_related_field, context={"request": serializer.context.get("request")}, **field_kwargs
             ).data
+
+
+class ModelAndFilterSetToUISchema:
+    """
+    Helper class to generate a JSON schema representation of a model or filterset, with fields mapped to UI widgets.
+
+    Args:
+        model (Model): The model to generate the schema for.
+        field_name (str): Optional. The name of a specific field to generate the schema for.
+        build_for (str): Optional. Generate the schema for a model form or a filter set form (default is "model_form").
+    """
+    def __init__(self, model, field_name=None, build_for="model_form"):
+        self.model = model
+        self.field_name = field_name
+        self.build_for = build_for
+        
+        # A mapping of Django Rest Framework serializer fields to their corresponding UI schemas.
+        self.serializer_field_mapping = {
+            serializers.CharField: self.char_field_schema,
+            serializers.IntegerField: self.integer_field_schema,
+            serializers.BooleanField: self.boolean_field_schema,
+            nautobot_serializers.NautobotPrimaryKeyRelatedField: self.dynamic_model_choice_field_schema,
+            serializers.ManyRelatedField: self.dynamic_model_multiple_choice_field_schema,
+            CustomFieldsDataField: self.custom_field_schema,
+            serializers.ChoiceField: self.choice_field_schema,
+            serializers.MultipleChoiceField: self.multiple_field_schema,
+            serializers.DateField: self.inheritance_schema_build,
+            serializers.DateTimeField: self.inheritance_schema_build,
+            serializers.TimeField: self.inheritance_schema_build,
+            serializers.JSONField: self.inheritance_schema_build,
+        }
+    
+    @property
+    def schema_representation(self):
```

Comment:
Given that this is a moderately complex calculation to generate, I'm not sure how I feel about marking it as a `property`. Might be better to be more explicit and keep it as a `get_schema_representation()` function instead.